NGSTACK-906 eztags priority sorting

AntePrkacin opened the pull request:
**Two main things:**
- added `priority` property to the Tag object
- added functionality to **sort** tags by either id, keyword, modified or priority

The sorting functionality is added below the children list for some tag. If tag has no children, then the sorting functionality is not rendered at all. 

The sorting works by modifying the `getChildren()` method in the Gateway layer. It first sorts and orders the fetched data and then slices it by using offset and limit values.

AntePrkacin committed:
NGSTACK-906 add 'priority' property to Tag objects

AntePrkacin committed:
NGSTACK-906 add support for 'priority' property in the Mapper layer

AntePrkacin committed:
NGSTACK-906 modify createTagFindQuery method to select 'priority' property

AntePrkacin committed:
NGSTACK-906 add 'priority' column to 'eztags' table in MySQL and PostgreSQL database schemas

AntePrkacin committed:
NGSTACK-906 add translations for tag priority

AntePrkacin committed:
NGSTACK-906 show 'priority' column for tag's children

AntePrkacin committed:
NGSTACK-906 add 'subitems' tab to the show tag template

AntePrkacin committed:
NGSTACK-906 add 'subitems' tab template with dropdowns for sortBy and sortOrder

AntePrkacin committed:
NGSTACK-906 add translations for tag subitems sorting

AntePrkacin committed:
NGSTACK-906 update tag children sorting functionality in TagController and update adapter implementation

AntePrkacin committed:
NGSTACK-906 add support for sorting in loadTagChildren() method in Service layer

AntePrkacin committed:
NGSTACK-906 add support for sorting in loadChildren() method for Tag Handler

AntePrkacin committed:
NGSTACK-906 add support for sorting in getChildren() method for Gateway layer

AntePrkacin committed:
NGSTACK-906 update subitems sort option value from 'tag_id' to 'id' in admin UI

AntePrkacin committed:
NGSTACK-906 remove 'subitems' tag tab and add the tag sorting at the bottom of children list

AntePrkacin committed:
NGSTACK-906 update tests to handle 'priority' property

emodric commented on bundle/API/Repository/TagsService.php:
```diff
@@ -71,12 +71,14 @@ public function loadTagByUrl(string $url, array $languages): Tag;
      * @return \Netgen\TagsBundle\API\Repository\Values\Tags\TagList
      */
-    public function loadTagChildren(?Tag $tag = null, int $offset = 0, int $limit = -1, ?array $languages = null, bool $useAlwaysAvailable = true): TagList;
+    public function loadTagChildren(?Tag $tag = null, int $offset = 0, int $limit = -1, ?array $languages = null, bool $useAlwaysAvailable = true, ?string $sortBy = null, ?string $sortOrder = null): TagList;
```
Do we need this level of control, in contrast to just using whatever info is storred on parent tag?

AntePrkacin replied:
I think this level of control is necessary because in the Gateway layer (`getChildren()` method), the fetched data from the database needs to **first be sorted** and then **sliced** using $limit and $offset values. 

Which _info stored on parent tag_ do you mean? Perhaps another approach for sorting would be if each Tag object would have information attached to it about the sortBy and sortOrder, something like the Location object in Ibexa has (if I'm not mistaken). 

EDIT: read the comment below this one. OK, I will implement the sorting mechanism similar to the Ibexa one

emodric commented on bundle/Controller/Admin/TagController.php:
```diff
@@ -45,6 +45,12 @@ public function showTagAction(Request $request, ?Tag $tag = null): Response
         if (!$tag instanceof Tag || !$tag->isSynonym()) {
             $configResolver = $this->getConfigResolver();
 
+            $sortBy = $request->query->get('sort_by');
```
I was under the impression that the sorting mechanism will be stored on parent tag, similar to how Ibexa does it in their locations. That way, you don't need to set the sorting for every children fetch, it will automatically fetch tags with the defined sort.

Default sort can then be set in configuration.

emodric commented on bundle/Core/Persistence/Legacy/Tags/Gateway/DoctrineDatabase.php:
```diff
@@ -137,10 +137,17 @@ public function getChildren(int $tagId, int $offset = 0, int $limit = -1, ?array
-        ->setMaxResults($limit > 0 ? $limit : PHP_INT_MAX);
+        )->setParameter('parent_id', $tagId, Types::INTEGER);
+
+        if ($sortBy !== null && $sortOrder !== null) {
```
We should have an enum with allowed sorting property and sorting directions, so we don't need to pass direct values. This way, we would only allow sorting by predefined columns, and not besically anything the user sends down the method call chain.

AntePrkacin replied:
Added enums for sortBy and sortOrder: https://github.com/netgen/TagsBundle/pull/170/commits/4224d33ccf899f460ac94d11cac29658ef811637
Also used enum cases to iterate over different options when selecting the property by which the user wants to sort the tags.

emodric commented on bundle/Core/Persistence/Legacy/Tags/Gateway/DoctrineDatabase.php:
```diff
@@ -137,10 +137,17 @@ public function getChildren(int $tagId, int $offset = 0, int $limit = -1, ?array
+            $tagIdsQuery->orderBy('eztags.keyword', 'ASC');
+        }
+
+        $tagIdsQuery
```
You don't need to separate `setFirstResult` and `setMaxResults` method calls from the main query builder calls.

AntePrkacin replied:
I put it back together, after refactoring the whole process of sorting children tags

AntePrkacin committed:
NGSTACK-906 add enums for tag sortField and sortOrder properties

AntePrkacin committed:
NGSTACK-906 add sortField and sortOrder properties to Tag objects and TagInfo object

AntePrkacin committed:
NGSTACK-906 update Mappers with sortField and sortOrder fields

AntePrkacin committed:
NGSTACK-906 remove sortBy and sortOrder parameters from the getChildren() method in Gateway layer

AntePrkacin committed:
NGSTACK-906 remove sortBy and sortOrder parameters from the loadChildren() method in Handler layer

AntePrkacin committed:
NGSTACK-906 remove sortBy and sortOrder parameters from the loadTagChildren() method in Service layer

AntePrkacin committed:
NGSTACK-906 rename subitems template to sort_children_tags template and move it out of tabs folder

AntePrkacin committed:
NGSTACK-906 add sortField and sortOrder properties to UpdateStruct and TagStruct classes

AntePrkacin committed:
NGSTACK-906 remove sorting logic in ChildrenTagsAdapter

AntePrkacin committed:
NGSTACK-906 add new route and controller method for updating sortField and sortOrder data for some tag

AntePrkacin committed:
NGSTACK-906 rename sortField property to sortBy property everywhere

AntePrkacin committed:
NGSTACK-906 add logic for sorting children by sortBy and sortOrder in DoctrineDatabase class

emodric commented on bundle/Controller/Admin/TagController.php:
```diff
@@ -270,6 +266,26 @@ public function updateTagAction(Request $request, Tag $tag, string $languageCode
+        $sortOrder = $request->request->get('sort_order');
+
+        $tagUpdateStruct = new TagUpdateStruct();
+        $tagUpdateStruct->sortField = TagSortField::from((string) $sortBy);
```
You can use `::tryFrom` here, catch the exception and throw `BadRequestHttpException`. Otherwise, transfering an invalid value would get you error 500.

AntePrkacin replied:
Did so here: https://github.com/netgen/TagsBundle/pull/170/commits/e16f73f5f80e6e2de49d5556ef63cec7e894c1c6

emodric commented on bundle/Core/Persistence/Legacy/Tags/Gateway/DoctrineDatabase.php:
```diff
@@ -128,6 +128,10 @@ public function getFullTagDataByKeywordAndParentId(string $keyword, int $parentI
     public function getChildren(int $tagId, int $offset = 0, int $limit = -1, ?array $translations = null, bool $useAlwaysAvailable = true): array
     {
+        $tagData = $tagId !== 0 ? $this->getBasicTagData($tagId) : [];
+        $sortBy = $tagData['sort_by'] ?? 'id';
```
We should add configuration for the default sort.

We also need configuration for the default sort for root tag (which has ID = 0), since we can't store this in the database.

emodric replied:
You can also try to instantiate the enum to double check that the value stored in the database is an allowed stored type and direction.

AntePrkacin replied:
First four commits made on 25 June 2025 are regarding the config parameters for default sorting.

There are **two new config parameters** added:
- one for default sorting of all child tags (`sort.by` and `sort.order`)
- one for default sorting of root tag children (`sort.root.by` and `sort.root.order`)

These config parameters are enums (enumNode), so their values will always match the cases of enum. There is even an [option to put an enum class in the node](https://symfony.com/doc/current/components/config/definition.html#enum-nodes), but that is only for Symfony 7.3 (I don't know which Symfony versions you want this bundle to support).

Also, in the database, the 'sort_by' and 'sort_order' column types were changed from enums to just varchar(100). By default, when a new tag is added, it will have ~empty string ('')~ NULL set at 'sort_by' and 'sort_order' columns - in this case, these two new config parameters will be fetched for sorting. 
Only when the user sorts the children tags of some tag, then will the 'sort_by' and 'sort_order' columns be set for that (parent) tag. 

This was done because the `<someEnum>::tryFrom()` method returns null if the corresponding (given) string has no matching case. In that case, the configResolver is used to get a parameter. Example:
```
$tag->sortBy = TagSortBy::tryFrom($row['sort_by'])
                        ?? TagSortBy::from($this->configResolver->getParameter('sort.by', 'netgen_tags'));
```

AntePrkacin committed:
NGSTACK-906 fix tests regarding sortBy and sortOrder properties

AntePrkacin committed:
NGSTACK-906 update TagSortBy and TagSortOrder enums, iterate through enum cases in sort_children_tags template and add translation for enums

AntePrkacin committed:
NGSTACK-906 throw BadRequestHttpException if sortBy or sortOrder params don't fit the enums when updating children sorting for some tag

AntePrkacin committed:
NGSTACK-906 change sort_by and sort_order column types from enums to varchar in database

AntePrkacin committed:
NGSTACK-906 add config params for sorting all child tags and for sorting specifically root tag children

AntePrkacin committed:
NGSTACK-906 update Mapper and DoctrineDatabase Gateway with logic for determining sortBy and sortOrder properties

AntePrkacin committed:
NGSTACK-906 add support for 'sort' config params in tests

AntePrkacin commented:
Should the dropdowns for sorting by and sorting order be added/shown to the 'Top level tags'? Or is it good enough to just be able to set the `sort.root.by` and `sort.root.order` config parameters for sorting the children of the 'Top level tags'?

AntePrkacin committed:
NGSTACK-906 add ability to modify 'priority' property when editing a tag and also show tag priority below the tag title

AntePrkacin committed:
NGSTACK-906 fix the update() method in DoctrineDatabase and the test for it

AntePrkacin commented:
https://github.com/netgen/TagsBundle/pull/170/commits/148a3c3edeba75fe3c560a7ed4669bc96edae44c
Added ability to change the `priority` property when editing a certain tag. Also, the information about tag priority will be displayed below the tag title (along with Tag ID, Remote ID and Modified properties).

emodric commented:
We don't have anywhere to store this info, so for now, the config is good enough for top level tags.

AntePrkacin committed:
NGSTACK-906 update 'sortBy' and 'sortOrder' properties to be null on Tag objects and in the database

AntePrkacin committed:
NGSTACK-906 update getChildren() and update() methods in DoctrineDatabase to properly handle the case when sortBy or sortOrder is null

AntePrkacin committed:
NGSTACK-906 update Mapper to properly handle the case when sortBy or sortOrder is null and extracted that logic in a separate function

AntePrkacin committed:
NGSTACK-906 update sort_children_tags template to check if sortBy or sortOrder is null and fix TagsIntegrationTest

AntePrkacin commented:
https://github.com/netgen/TagsBundle/pull/170/commits/59beb45a538d7bdbc0523b72afb82741249a4d56 https://github.com/netgen/TagsBundle/pull/170/commits/84d9758c6e0c75d678bad7ee1f7b18118d790c03 https://github.com/netgen/TagsBundle/pull/170/commits/fd507ddc4873c3335ce1774a99ef067c064c756c https://github.com/netgen/TagsBundle/pull/170/commits/a81d3f6e68f79483b012fa389767160b043a726e 
There was a problem when editing a certain tag. For example, when you went to edit a priority for some tag and hit 'save', the `sort_by` and `sort_order` properties (and columns in the database) would automatically be assigned the default values (from the config params). 

I wanted to avoid this because then, if someone changes the config params for default sorting, the sorting will not change for this one tag because (unintentionally) editing the tag already set the sorting. 

With the commits I pasted above, this problem is handled by making the `sortBy` and `sortOrder` properties be null (by default - both for the database and for the entity). ONLY when the user clicks to sort the tags (the thing the user can click below the children list in adminUI), only then are the `sort_by` and `sort_order` columns set in the database.

AntePrkacin committed:
NGSTACK-906 add MySQL and PostgreSQL upgrade scripts for 'priority', 'sort_by' and 'sort_order' columns and updated the UPGRADE doc

AntePrkacin commented:
https://github.com/netgen/TagsBundle/pull/170/commits/0ce8c25a01e0b1b9974b82ba44e2a9e2795a8c17
Updated the `doc/UPGRADE.md` file with update instructions for this PR. Also added the upgrade scripts to add `priority`, `sort_by` and `sort_order` columns to `eztags` table. 

https://github.com/netgen/TagsBundle/pull/170/commits/3e9628a82fdf0ea5914d0d795d74f0fdb3c0c7b0
Removed 'priority' info when showing synonyms and removed the 'priority' field when editing or adding a synonym to some tag. I think it makes no sense to show 'priority' info on synonyms, since there is no sorting for them at all. These synonyms will still have priority be 0 in the database, but that data won't be shown anywhere.

https://github.com/netgen/TagsBundle/pull/170/commits/031c1aa1d3f6bc67f99764195bf8f64d2ec84d12 
Added policy for sorting

AntePrkacin committed:
NGSTACK-906 don't show 'priority' info on synonyms and remove 'priority' field when adding or editing a synonym

AntePrkacin committed:
NGSTACK-906 add 'priority', 'sort_by' and 'sort_order' fields to legacy schema

AntePrkacin committed:
NGSTACK-906 add policy for sorting children tags

pspanja commented:
@AntePrkacin 
> Should the dropdowns for sorting by and sorting order be added/shown to the 'Top level tags'? Or is it good enough to just be able to set the sort.root.by and sort.root.order config parameters for sorting the children of the 'Top level tags'?

From my POV we do need it on "Top-level tags" as well, editors need a way to sort its children, and being able to do it through configuration only means they would need a developer in order to change it.
